src/Wasmtime.csproj

```diff
@@ -37,6 +37,34 @@ The .NET embedding of Wasmtime enables .NET code to instantiate WebAssembly modu
   <ItemGroup>
     <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.1.1" PrivateAssets="All" />
   </ItemGroup>
+  
+  <ItemGroup>
+    <None Update="Linker.DefineFunction.tt">
+      <Generator>TextTemplatingFileGenerator</Generator>
+      <LastGenOutput>Linker.DefineFunction.cs</LastGenOutput>
+    </None>
+    <Compile Update="Linker.DefineFunction.cs">
+      <DesignTime>True</DesignTime>
+      <AutoGen>True</AutoGen>
+      <DependentUpon>Linker.DefineFunction.tt</DependentUpon>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="Function.FromCallback.tt">
+      <Generator>TextTemplatingFileGenerator</Generator>
+      <LastGenOutput>Function.FromCallback.cs</LastGenOutput>
+    </None>
+    <Compile Update="Function.FromCallback.cs">
+      <DesignTime>True</DesignTime>
+      <AutoGen>True</AutoGen>
+      <DependentUpon>Function.FromCallback.tt</DependentUpon>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup>
+    <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
```

Comment:
I'm wondering if instead of relying on a VS service to do the code generation, we install the `dotnet-t4` tool using a [local tool manifest](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-local-tool) and invoke it via a custom "generate" target that the build target depends on?

That way it should work both for VS and for the .NET CLI (provided users runs `dotnet tool restore` prior to building).

What are your thoughts on this approach?

I'd also like, if possible, to not add the generated code to the repository; it's easy enough to inspect locally, if needed, and the version history on the template is more important than on the generated file.

Reply:
Thanks, good idea! I added the tool manifest file and added commands to the project file, so that the `dotnet-t4` tool is automatically run on each build. I also added the `dotnet tool restore` command to automatically make it available (as otherwise it would fail e.g. on CI). See also https://github.com/dotnet/sdk/issues/4241.

> I'd also like, if possible, to not add the generated code to the repository; it's easy enough to inspect locally, if needed, and the version history on the template is more important than on the generated file.

Theoretically it would now be possible to omit these files from source control, because they are now generated on each build. However, AFAIK this would have some drawbacks:
- When doing a fresh clone of the repo (or doing a clean-up) and then opening the solution in an IDE, users would first need to do a build in order to generate these files. Otherwise, users might get errors or wrong overloads being shown, like this in the `FunctionTests`: 
![grafik](https://user-images.githubusercontent.com/13289184/199814800-ec95a8bf-5055-41ea-92ae-10bb73b081be.png)<br>
  Only after the build, the correct overloads will be shown:
![grafik](https://user-images.githubusercontent.com/13289184/199814963-260bdc11-9b2d-4ee4-8b09-02b141e860a2.png)<br><br>
- I'm not sure whether SourceLink will work correctly for the generated files, because as far as I understand it, it will download the source code file from Github for a specific commit. However, when these files would not be in the repository, I think showing the source code would fail.

What do you think?
Thanks!

Reply:
I'm not as concerned about a poor developer experience until a recently-cloned project is built for the first time, but that's a really good point about not including the generating files will probably impact source link.

Let's keep the generated files for now as it's not really hurting things having them in the repo.

Also no problem keeping the `Service` entry in the project file as it is entirely ignored by the CLI tools.